File: 1871_Jump_Game_VII/solution.c
bool canReach(char* s, int minJump, int maxJump) {
    int n = strlen(s);
    int* f = (int*)calloc(n, sizeof(int));
    int* p = (int*)malloc(n * sizeof(int));
    f[0] = 1;
    for (int i = 0; i < minJump; i++) p[i] = 1;
    for (int i = minJump; i < n; i++) {
        int l = i - maxJump, r = i - minJump;
        if (s[i] == '0') {
            int t = p[r];
            if (l > 0) t -= p[l - 1];
            f[i] = t != 0 ? 1 : 0;
        }
        p[i] = p[i - 1] + f[i];
    }
    bool r = (f[n - 1] == 1);
    free(f);
    free(p);
    return r;
}

File: 1871_Jump_Game_VII/solution.py
class Solution:
    def canReach(self, s: str, mj: int, maxJump: int) -> bool:
        n = len(s)
        f, p = [0] * n, [0] * n
        f[0] = 1
        for i in range(mj):
            p[i] = 1
        for i in range(mj, n):
            l, r = i - maxJump, i - mj
            if s[i] == "0":
                total = p[r] - (0 if l <= 0 else p[l - 1])
                f[i] = int(total != 0)
            p[i] = p[i - 1] + f[i]

        return bool(f[n - 1])
